src/main/java/io/github/rhkiswani/javaff/httpclient/ApacheHttpClient2.java

```diff
@@ -0,0 +1,181 @@
+ *
+ */
+public class ApacheHttpClient2 implements HttpClient{
+
+    @Override
+    public String postJson(String url, String json, Map<String, String> headers) throws HttpClientException {
+        return doJsonRequest(new HttpPost(url), json, headers);
+    }
+
+    @Override
+    public String putJson(String url, String json, Map<String, String> headers) throws HttpClientException {
+        return doJsonRequest(new HttpPut(url), json, headers);
+    }
+
+    private String doJsonRequest(HttpRequestBase method, String json, Map<String, String> headers) {
+        try {
+            if (json == null){
+                throw new HttpClientException(SmartException.NULL_VAL, "Json Object");
+            }
+            CloseableHttpClient c =  HttpClientBuilder.create().build();
+            ((HttpEntityEnclosingRequestBase) method).setEntity(new StringEntity(json));
+            method.setHeader("Accept", "application/json");
+            method.setHeader("Content-type", "application/json");
+            setHeaders(method, headers);
+            return handleResponse(method, c);
+        } catch (Exception e) {
+            throw new HttpClientException(e.getMessage());
+        }
+    }
+
+    private String handleResponse(HttpRequestBase method, CloseableHttpClient c) {
+        HashMap<String, Object> hashMap = new HashMap<>();
+        try {
+            HttpResponse response = c.execute(method);
+            hashMap.put("response", IOUtil.inputStreamToString(response.getEntity().getContent()));
+            hashMap.put("status", response.getStatusLine().getStatusCode());
+            hashMap.put("headers", IOUtil.headersToMap(response.getAllHeaders()));
+        } catch (Exception ex) {
+            hashMap.put("status", HttpStatus.SC_BAD_REQUEST );
+            hashMap.put("errors", ReflectionUtil.getErrorsAsArray(ex));
+        }
+        return JsonHandlerFactory.getJsonHandler(HashMap.class).toJson(hashMap);
+    }
+
+    @Override
+    public String post(String url, Map<String, String> params, Map<String, String> headers) throws HttpClientException {
+        return doRequest(new HttpPost(url), params, headers);
+    }
+
+    @Override
+    public String put(String url, Map<String, String> params, Map<String, String> headers) throws HttpClientException {
+        return doRequest(new HttpPut(url), params, headers);
+    }
+
+    private String doRequest(HttpRequestBase method, Map<String, String> params, Map<String, String> headers) {
+        try {
+            CloseableHttpClient c = prepareRequest(method, params);
+            setHeaders(method, headers);
+            return handleResponse(method, c);
+        } catch (Exception e) {
+            throw new HttpClientException(e.getMessage());
+        }
+    }
+
+    private CloseableHttpClient prepareRequest(HttpRequestBase method, Map<String, String> params) throws UnsupportedEncodingException {
+        CloseableHttpClient c =  HttpClientBuilder.create().build();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *c*. The resource is returned at line: 119. There are other execution paths that neither close the resource nor return it, e.g., when exception is thrown by BasicNameValuePair constructor, and if-branch is taken at line 109. Either a) close *c* in a try-finally block or b) close the resource by declaring *c* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).